pkg/distributor/push_test.go

```diff
@@ -872,3 +881,199 @@ func TestRetryConfig_Validate(t *testing.T) {
 		})
 	}
 }
+
+func TestOTLPPushHandlerErrorsAreReportedCorrectlyViaHttpgrpc(t *testing.T) {
+	reg := prometheus.NewRegistry()
+	cfg := dskit_server.Config{}
+	// Set default values
+	cfg.RegisterFlags(flag.NewFlagSet("test", flag.ContinueOnError))
+
+	// Configure values for test.
+	cfg.HTTPListenAddress = "localhost"
+	cfg.HTTPListenPort = 0 // auto-assign
+	cfg.GRPCListenAddress = "localhost"
+	cfg.GRPCListenPort = 0 // auto-assign
+	cfg.Registerer = reg
+	cfg.Gatherer = reg
+	cfg.ReportHTTP4XXCodesInInstrumentationLabel = true // report 400 as errors.
+	cfg.GRPCMiddleware = []grpc.UnaryServerInterceptor{middleware.ServerUserHeaderInterceptor}
+	cfg.HTTPMiddleware = []middleware.Interface{middleware.AuthenticateUser}
+
+	srv, err := dskit_server.New(cfg)
+	require.NoError(t, err)
+	//srv.HTTP.
+
+	push := func(ctx context.Context, req *Request) error {
+		// Trigger conversion of incoming request to WriteRequest.
+		wr, err := req.WriteRequest()
+		if err != nil {
+			return err
+		}
+
+		if len(wr.Timeseries) > 0 && len(wr.Timeseries[0].Labels) > 0 && wr.Timeseries[0].Labels[0].Name == "__name__" && wr.Timeseries[0].Labels[0].Value == "report_server_error" {
+			return errors.New("some random push error")
+		}
+
+		return nil
+	}
+	h := OTLPHandler(200, util.NewBufferPool(), nil, false, otlpLimitsMock{}, RetryConfig{Enabled: false}, push, newPushMetrics(reg), reg, log.NewNopLogger(), true)
+	srv.HTTP.Handle("/otlp", h)
+
+	// start the server
+	require.NoError(t, err)
+	go func() { _ = srv.Run() }()
```

Comment:
Could you use a `sync.WaitGroup` in the goroutine here, just so the test can wait for the goroutine to stop? I've had race conditions in other tests because of non-deterministic goroutines, which were easily fixed in this way.

Reply:
Good idea, done in https://github.com/grafana/mimir/pull/8363/commits/153324c1accb20719f83eb8db8b0adc3dd9798b8.